### ISSUE TITLE
Can't cancel NextTweetStreamAsync

### ISSUE THREAD

**Issue by marcogruhl** [Xwilarg/TwitterSharp#15]
when i use `NextTweetStreamAsync` the way it is described in the examples i can not stop/cancel it. It results in a "Too many Streams error" from twitter.

```cs
Task.Run(async () =>
{
    // Take in parameter a callback called for each new tweet
    // Since we want to get the basic info of the tweet author, we add an empty array of UserOption
    await client.NextTweetStreamAsync((tweet) =>
    {
        Console.WriteLine($"From {tweet.Author.Name}: {tweet.Text}");
    }, null, Array.Empty<UserOption>());
});
```
when i add a cancellationToken i can control it:

```cs
public async Task NextTweetStreamAsync(Action<Tweet> onNextTweet, CancellationToken cancellationToken, TweetOption[] tweetOptions = null, UserOption[] options = null, MediaOption[] mediaOptions = null)
{
    var req = new RequestOptions();
    AddTweetOptions(req, tweetOptions);
    AddUserOptions(req, options, true);
    AddMediaOptions(req, mediaOptions);
    var res = await _httpClient.GetAsync(_baseUrl + "tweets/search/stream?" + req.Build(), HttpCompletionOption.ResponseHeadersRead, cancellationToken);
    BuildRateLimit(res.Headers, "NextTweetStreamAsync");
    using StreamReader reader = new(await res.Content.ReadAsStreamAsync(cancellationToken));
    while (!reader.EndOfStream && !cancellationToken.IsCancellationRequested)
    {
        var str = reader.ReadLine();
        if (string.IsNullOrWhiteSpace(str))
        {
            continue;
        }
        onNextTweet(ParseData<Tweet>(str).Data);
    }
}
```
I will investigate this further and create a pull request if i find a suitable solution. Maybe a general Dispose function to avoid the specific cancellationToken and prevent breaking changes.

Am i doing something wrong?

Any thoughts or ideas?

**Comment by Xwilarg** [comment]
Hey, thanks for the bug report, as if I'm not sure what the issue could be but I'll see if I can reproduce it and find a fix
Meanwhile if you find the solution, feel free to share it :)

**Comment by marcogruhl** [comment]
I created a branch with my approach: [Dispose Client](https://github.com/marcogruhl/TwitterSharp/tree/ClientDispose)
It has no breaking changes and all tests run successfully.

**Comment by Xwilarg** [comment]
Looks good but what happens if the user call NextTweetStreamAsync twice?

**Comment by marcogruhl** [comment]
> Looks good but what happens if the user call NextTweetStreamAsync twice?

I am not sure if this is even a wanted behavior:

- _"No disconnection needed to add/remove rules using rules endpoint."_  ([Filtered Stream Endpoints overview](https://developer.twitter.com/en/docs/twitter-api/filtered-stream-overview))
- The redundancy feature looks like it is supposed to be used with independent HttpClients or even redundant Systems: _"A redundant connection simply allows you to establish more than one simultaneous connections to the filtered stream. This provides redundancy by allowing you to connect to the same stream with two separate consumers, receiving the same data through both connections. Thus, your app has a hot failover for various situations such as if one stream is disconnected or if your application's primary server fails."_  
and it is only available to academic access _"Filtered stream currently only allows Projects with Academic Research access to connect to up to two redundant connections. To use a redundant stream, simply connect to the same URL used for your primary connection. The data for your stream will be sent through both connections."_ ([Filtered Stream Endpoints overview](https://developer.twitter.com/en/docs/twitter-api/filtered-stream-overview))

If this is the case, i can:
- implement some kind of "singleton" logic which returns to multiple actions from a single stream. With this "solution" the problem is to handle different options-parameter
- cancel and re-request the stream with new parameter. Vice versa we had the problem with the old actions and potential diffenrent options

If we want to allow multiple streams with one httpClient i can think of storing an array of cancallation tokens and Actions for cancallation.

**Comment by Xwilarg** [comment]
Uhm yeah, if you could at least add a small message in the XML comment on top of the function I think it could be nice, maybe using the <remarks> tag or something so it stand-out a bit?

**Comment by marcogruhl** [comment]
Closed with pull request #16
